File: Valid Parentheses/ValidParentheses.java
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> parentheses = new Stack<>();
        
        if (s.charAt(0) == ')' || s.charAt(0) == ']' || s.charAt(0) == '}') {
            return false;
        }

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            
            if (c == '(' || c == '[' || c == '{') {
                parentheses.push(c);
            } else if (c == ')') {
                if (parentheses.isEmpty() || parentheses.peek() != '(') {
                    return false;
                }
                parentheses.pop();
            } else if (c == ']') {
                if (parentheses.isEmpty() || parentheses.peek() != '[') {
                    return false;
                }
                parentheses.pop();
            } else if (c == '}') {
                if (parentheses.isEmpty() || parentheses.peek() != '{') {
                    return false;
                }
                parentheses.pop();
            }
        }

        return parentheses.isEmpty();
    }
}

File: Valid Parentheses/ValidParentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        
        parentheses = []
        
        if s[0] == ")" or s[0] == "]" or s[0] == "}":
            return False

        for i in range(len(s)):
            
            if s[i] == "(" or s[i] == "[" or s[i] == "{":
                parentheses.append(s[i])
            
            elif s[i] == ")":
                if len(parentheses) == 0:
                    return False
                if parentheses[-1] == "(":
                    parentheses.pop()
                    continue
                else:
                    return False

            elif s[i] == "]":
                if len(parentheses) == 0:
                    return False
                if parentheses[-1] == "[":
                    parentheses.pop()
                    continue
                else:
                    return False

            elif s[i] == "}":
                if len(parentheses) == 0:
                    return False
                if parentheses[-1] == "{":
                    parentheses.pop()
                    continue
                else:
                    return False

        if len(parentheses) > 0:
            return False

        return True